README.md
# bryt-lite

Get colors by brightness.

[![NPM Version][npm-image]][npm-url]
[![NPM Downloads][downloads-image]][downloads-url]

## Overview

bryt provides an API for getting all colors for a given brightness. A brightness is represented by
an integer between `0` and `255`. Some brightnesses have more colors than others.

Determining a color's brightness is a pretty simple calculation, but choosing a color by brightness
is not. bryt-lite uses a pre-baked lookup table of brightnesses and their associated colors.

There are 16,777,216 possible 24-bit colors which takes up 66 MB. With some clever hacks and Brotli
compression, we can squeeze this size down to a little over 3 MB, but this is still too big.

bryt-lite has reduced the number of colors per brightness that are similar to each other. Instead
of 16 million colors, bryt-lite only has 7,204 colors and requires 66 KB on disk.

If you want the full lookup table and don't mind the 3MB size, you can always use [bryt][bryt]. The
bryt-lite API is identical to bryt's API.

## Installation

    npm install bryt-lite

## Examples

Get brightness info.

```js
import * as bryt from 'bryt-lite';

const info = bryt.getBrightness(128);
console.log(`Brightness ${info.brightness} has ${info.count} colors`);

for (let i = 0; i < info.count; i++) {
	console.log(`Color ${i + 1}) ${info.getColor(i)}`);
}

console.log('All colors:', info.getColors());
```

Get a specific color by brightness and index.

```js
const color = bryt.getColor(200, 3);
```

Get all colors for a specific brightness. Note that this is not super performant.

```js
const colors = bryt.getColors(187);
for (const color of colors) {
	console.log(color);
}
```

Convert a integer color to an RGB array.

```js
const color = bryt.getColor(200, 3);
const [ red, green, blue ] = bryt.toRGB(color);
console.log(`red: ${red}, green: ${green}, blue ${blue}`);
```

## API

### `getBrightness(brightness)`

 * `brightness` (Number): A positive integer between 0 and 255.

Returns `Object` containing the `brightness`, `count` of colors, `getColor(idx)`, and `getColors()`.

### `getColor(brightness, idx)`

 * `brightness` (Number): A positive integer between 0 and 255.

Returns `Number` as a positive integer.

### `getColors(brightness)`

 * `brightness` (Number): A positive integer between 0 and 255.

Returns `Array<Number>` containing all colors (as integers).

### `toRGB(num)`

 * `num` (Number): A positive integer to split into red, green, and blue components.

Returns `Array<Number>`.

## License

MIT

[bryt]: https://npmjs.org/package/bryt
[npm-image]: https://img.shields.io/npm/v/bryt-lite.svg
[npm-url]: https://npmjs.org/package/bryt-lite
[downloads-image]: https://img.shields.io/npm/dm/bryt-lite.svg
[downloads-url]: https://npmjs.org/package/bryt-lite


CHANGELOG.md
# v2.0.1 (Jul 7, 2022)

 * chore: Updated dev dependencies.

# v2.0.0 (Mar 1, 2022)

 * BREAKING CHANGE: Require Node.js 14.15.0 LTS or newer.
 * chore: Updated dev dependencies.
 * chore: Replaced Travis with GitHub action.

# v1.0.1 (Feb 23, 2021)

 * chore: Added AppVeyor CI builds.
 * chore: Lint tests.
 * doc: Added changelog.

# v1.0.0 (Feb 19, 2021)

 * Initial release.


test/test-bryt.js
import * as bryt from '../src/index.js';
import { expect } from 'chai';

describe('getBrightness()', () => {
	it('should error if brightness is not an integer', () => {
		expect(() => {
			bryt.getBrightness('foo');
		}).to.throw(TypeError, 'Expected brightness to be an integer');

		expect(() => {
			bryt.getBrightness(123.45);
		}).to.throw(TypeError, 'Expected brightness to be an integer');
	});

	it('should error if brightness not in the valid range', () => {
		expect(() => {
			bryt.getBrightness(-10);
		}).to.throw(RangeError, 'Expected brightness to be between 0 and 255');

		expect(() => {
			bryt.getBrightness(1000);
		}).to.throw(RangeError, 'Expected brightness to be between 0 and 255');
	});

	it('should get info for a specific brightness', () => {
		const info = bryt.getBrightness(128);
		expect(info).to.be.an('object');
		expect(info).to.have.property('brightness', 128);
		expect(info).to.have.property('count', 51);

		expect(info.getColor(10)).to.equal(16728193);
		expect(() => {
			info.getColor('foo');
		}).to.throw(TypeError, 'Expected index to be an integer');
		expect(() => {
			info.getColor(-123);
		}).to.throw(RangeError, 'Expected index to be a positive integer');

		const colors = info.getColors();
		expect(colors).to.be.an('array');
		expect(colors).to.have.lengthOf(51);
	});
});

describe('getColor()', () => {
	it('should error if brightness is not a number', () => {
		expect(() => {
			bryt.getColor('foo');
		}).to.throw(TypeError, 'Expected brightness to be an integer');

		expect(() => {
			bryt.getColor(123.45);
		}).to.throw(TypeError, 'Expected brightness to be an integer');
	});

	it('should error if brightness not in the valid range', () => {
		expect(() => {
			bryt.getColor(-10);
		}).to.throw(RangeError, 'Expected brightness to be between 0 and 255');

		expect(() => {
			bryt.getColor(1000);
		}).to.throw(RangeError, 'Expected brightness to be between 0 and 255');
	});

	it('should error if index is not an integer', () => {
		expect(() => {
			bryt.getColor(0, 'foo');
		}).to.throw(TypeError, 'Expected index to be an integer');

		expect(() => {
			bryt.getColor(0, 123.45);
		}).to.throw(TypeError, 'Expected index to be an integer');
	});

	it('should error if index is negative', () => {
		expect(() => {
			bryt.getColor(0, -19);
		}).to.throw(RangeError, 'Expected index to be a positive integer');
	});

	it('should get a color', () => {
		expect(bryt.getColor(0, 0)).to.equal(0);
		expect(bryt.getColor(128, 10)).to.equal(16728193);
	});
});

describe('getColors()', () => {
	it('should error if brightness is not a number', () => {
		expect(() => {
			bryt.getColors('foo');
		}).to.throw(TypeError, 'Expected brightness to be an integer');

		expect(() => {
			bryt.getColors(123.45);
		}).to.throw(TypeError, 'Expected brightness to be an integer');
	});

	it('should error if brightness not in the valid range', () => {
		expect(() => {
			bryt.getColors(-10);
		}).to.throw(RangeError, 'Expected brightness to be between 0 and 255');

		expect(() => {
			bryt.getColors(1000);
		}).to.throw(RangeError, 'Expected brightness to be between 0 and 255');
	});

	it('should get all colors for a brightness', () => {
		const colors = bryt.getColors(200);
		expect(colors).to.be.an('array');
		expect(colors).to.have.lengthOf(40);
	});
});

describe('toRGB()', () => {
	it('should error if the color is invalid', () => {
		expect(() => {
			bryt.toRGB('foo');
		}).to.throw(TypeError, 'Expected color to be an integer');

		expect(() => {
			bryt.toRGB(123.45);
		}).to.throw(TypeError, 'Expected color to be an integer');
	});

	it('should error if the color is negative', () => {
		expect(() => {
			bryt.toRGB(-10);
		}).to.throw(RangeError, 'Expected color to be a positive integer');
	});

	it('should convert a color to RGB', () => {
		expect(bryt.toRGB(123)).to.deep.equal([ 0, 0, 123 ]);
		expect(bryt.toRGB(64326)).to.deep.equal([ 0, 251, 70 ]);
		expect(bryt.toRGB(1468399)).to.deep.equal([ 22, 103, 239 ]);
	});
});


scripts/build.js
import ansiColors from 'ansi-colors';
import fs from 'fs';
import os from 'os';
import path from 'path';
import { fileURLToPath } from 'url';
import { isMainThread, parentPort, Worker } from 'worker_threads';

/**
 * The color threshold is used to control how many colors are removed.
 *
 *    Threshold | Perception
 *   -----------+---------------------------------------
 *    <= 1.0    | Not perceptible by human eye
 *    1 - 2     | Perceptible through close observation
 *    2 - 10    | Perceptible at a glance
 *    11 - 49   | Colors are more similar than opposite
 *    100       | Colors are exact opposite
 *
 * The idea is to tune the threshold such that we have a good set of colors to choose from per
 * brightness without using a ton of space.
 *
 *   Threshold | Colors | Size (bytes) | Size on Disk
 *  -----------+--------+--------------+----------------
 *       5     | 7,204  | 65,055       | 66 KB
 *       6     | 5,103  | 46,167       | 49 KB
 *       7     | 3,555  | 32,376       | 33 KB
 *       8     | 2,771  | 25,529       | 29 KB
 */
const colorThreshold = 5;

const dirname = fileURLToPath(new URL('.', import.meta.url));
const cacheDir = path.resolve(dirname, '..', '.cache');

if (isMainThread) {
	const start = Date.now();
	const { cyan } = ansiColors;
	const index = [];
	const table = [];
	let brightness;

	if (!fs.existsSync(cacheDir)) {
		console.log('Calculating table...');
		for (let b = 0; b < 256; b++) {
			for (let g = 0; g < 256; g++) {
				for (let r = 0; r < 256; r++) {
					brightness = ((r * 299 + g * 587 + b * 114) / 1000) | 0;
					if (!table[brightness]) {
						table[brightness] = [];
					}
					table[brightness].push([
						rgb2lab(r, g, b),
						(r << 16) + (g << 8) + b
					]);
				}
			}
		}

		console.log('Writing cache...');
		fs.mkdirSync(cacheDir);

		// we need to manually write the cache file as it's too big to JSON.stringify()
		for (let i = 0; i < 256; i++) {
			fs.writeFileSync(path.join(cacheDir, `${i}.json`), JSON.stringify(table[i].sort((a, b) => a[1] - b[1])));
		}
	}

	brightness = 0;
	const numWorkers = os.cpus().length / 2;
	const workers = [];
	let totalColors = 0;

	console.log(`Starting ${cyan(numWorkers)} workers...`);
	for (let i = 0; i < numWorkers; i++) {
		const w = new Worker(fileURLToPath(import.meta.url));
		w.id = i;
		w.on('message', msg => {
			if (msg) {
				const { brightness, colors, count } = msg;
				index[brightness] = colors;
				totalColors += colors.length;
				console.log(
					`Worker: ${cyan(String(w.id).padEnd(2))} ` +
					`Brightness: ${cyan(String(brightness).padEnd(3))} ` +
					`Total Count: ${cyan(String(count).padEnd(7))} ` +
					`Removed: ${cyan(String(count - colors.length).padEnd(7))} ` +
					`Colors: ${cyan(String(colors.length).padEnd(7))}`
				);
			}
			if (brightness < 256) {
				w.postMessage({ brightness: brightness++ });
			} else {
				w.terminate();
			}
		});
		workers.push(new Promise((resolve, reject) => w.on('error', reject).on('exit', resolve)));
	}

	await Promise.all(workers);

	const lookupFile = path.resolve(dirname, '../src/lookup.js');
	let s = 'export const lookup=[';
	for (let i = 0, l = index.length; i < l; i++) {
		const colors = index[i];
		if (i) {
			s += ',';
		}
		s += '[';
		for (let j = 0, k = colors.length; j < k; j++) {
			if (j) {
				s += ',';
			}
			s += colors[j];
		}
		s += ']';
	}
	s += '];';
	fs.writeFileSync(lookupFile, s, 'utf-8');

	console.log(`Finished in ${cyan(Date.now() - start)} ms`);
	console.log(`Total colors: ${cyan(totalColors)}`);
	console.log(`Lookup file size: ${cyan(fs.statSync(lookupFile).size)} bytes`);
} else {
	// worker thread
	parentPort.on('message', async ({ brightness }) => {
		let colors = JSON.parse(fs.readFileSync(path.join(cacheDir, `${brightness}.json`)));
		const count = colors.length;

		// remove similar colors
		for (let i = 0; i < colors.length; i++) {
			const result = [ colors[i] ];
			for (let j = i + 1; j < colors.length; j++) {
				if (deltaE(colors[i][0], colors[j][0]) >= colorThreshold) {
					result.push(colors[j]);
				}
			}
			colors = result;
		}

		parentPort.postMessage({
			brightness,
			colors: colors.map(c => c[1]),
			count
		});
	});

	parentPort.postMessage(null);
}

// https://stackoverflow.com/a/52453462/5501041
function rgb2lab(r, g, b) {
	let x, y, z;
	r /= 255;
	g /= 255;
	b /= 255;
	r = (r > 0.04045) ? Math.pow((r + 0.055) / 1.055, 2.4) : r / 12.92;
	g = (g > 0.04045) ? Math.pow((g + 0.055) / 1.055, 2.4) : g / 12.92;
	b = (b > 0.04045) ? Math.pow((b + 0.055) / 1.055, 2.4) : b / 12.92;
	x = (r * 0.4124 + g * 0.3576 + b * 0.1805) / 0.95047;
	y = (r * 0.2126 + g * 0.7152 + b * 0.0722) / 1.00000;
	z = (r * 0.0193 + g * 0.1192 + b * 0.9505) / 1.08883;
	x = (x > 0.008856) ? Math.pow(x, 1 / 3) : (7.787 * x) + 16 / 116;
	y = (y > 0.008856) ? Math.pow(y, 1 / 3) : (7.787 * y) + 16 / 116;
	z = (z > 0.008856) ? Math.pow(z, 1 / 3) : (7.787 * z) + 16 / 116;
	return [ (116 * y) - 16, 500 * (x - y), 200 * (y - z) ];
}

function deltaE(labA, labB) {
	const deltaL = labA[0] - labB[0];
	const deltaA = labA[1] - labB[1];
	const deltaB = labA[2] - labB[2];
	const c1 = Math.sqrt(labA[1] * labA[1] + labA[2] * labA[2]);
	const c2 = Math.sqrt(labB[1] * labB[1] + labB[2] * labB[2]);
	const deltaC = c1 - c2;
	let deltaH = deltaA * deltaA + deltaB * deltaB - deltaC * deltaC;
	deltaH = deltaH < 0 ? 0 : Math.sqrt(deltaH);
	const sc = 1.0 + 0.045 * c1;
	const sh = 1.0 + 0.015 * c1;
	const deltaLKlsl = deltaL / (1.0);
	const deltaCkcsc = deltaC / (sc);
	const deltaHkhsh = deltaH / (sh);
	const i = deltaLKlsl * deltaLKlsl + deltaCkcsc * deltaCkcsc + deltaHkhsh * deltaHkhsh;
	return i < 0 ? 0 : Math.sqrt(i);
}


src/index.js
import { lookup } from './lookup.js';

/**
 * Checks that the brightness value is valid.
 *
 * @param {Number} brightness - The brightness level.
 */
function assertBrightness(brightness) {
	if (typeof brightness !== 'number' || ~~brightness !== brightness) {
		throw new TypeError('Expected brightness to be an integer');
	}

	if (brightness < 0 || brightness > 255) {
		throw new RangeError('Expected brightness to be between 0 and 255');
	}
}

/**
 * Gets information about a specific brightness and scoped functions for retreiving that brightness
 * colors or a specific color by index.
 *
 * @param {Number} brightness - The brightness level.
 * @returns {Object}
 */
export function getBrightness(brightness) {
	assertBrightness(brightness);

	return {
		brightness,
		count: lookup[brightness].length,
		getColor(idx) {
			return getColor(brightness, idx);
		},
		getColors() {
			return getColors(brightness);
		}
	};
}

/**
 * Returns a color at the specified index for the given brightness.
 *
 * @param {Number} brightness - The brightness level.
 * @param {Number} idx - The index in the list of colors.
 * @returns {Number}
 */
export function getColor(brightness, idx) {
	assertBrightness(brightness);

	if (typeof idx !== 'number' || ~~idx !== idx) {
		throw new TypeError('Expected index to be an integer');
	}

	if (idx < 0) {
		throw new RangeError('Expected index to be a positive integer');
	}

	return lookup[brightness][idx];
}

/**
 * Returns all colors for the specified brightness.
 *
 * @param {Number} brightness - The brightness level.
 * @returns {Array.<Number>}
 */
export function getColors(brightness) {
	assertBrightness(brightness);
	return lookup[brightness];
}

/**
 * Converts a number to an array of red, green, and blue component values.
 *
 * @param {Number} num - A number representing a color.
 * @returns {Array.<Number>}
 */
export function toRGB(num) {
	if (typeof num !== 'number' || ~~num !== num) {
		throw new TypeError('Expected color to be an integer');
	}

	if (num < 0) {
		throw new RangeError('Expected color to be a positive integer');
	}

	return [ (num >> 16) & 0xFF, (num >> 8) & 0xFF, num & 0xFF ];
}
